bug: AUR build is failing

Trying to install the new release it failing:


```bash
   Compiling dalfox v3.0.1 (/home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1)
error: linking with `x86_64-linux-gnu-gcc` failed: exit status: 1
  |
  = note:  "x86_64-linux-gnu-gcc" "-m64" "<1 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/rustcWQELDm/libaws_lc_sys-d947a2f413365547.rlib" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-*.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-L" "/home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/rustcWQELDm/raw-dylibs" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/build/aws-lc-sys-9342da2c99d4592a/out" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-Wl,--strip-all" "-nodefaultlibs"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_PKEY_bits
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_digest_sig::hf138b2418df46f53)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::ec::key_pair::EcdsaKeyPair::new::hcdb47f2e2bead0c0)
          >>> referenced 2 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_MD_CTX_init
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced 5 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_DigestVerifyInit
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced 1 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_PKEY_CTX_set_rsa_padding
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_digest_sig::hf138b2418df46f53)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$rustls..crypto..aws_lc_rs..sign..RsaSigner$u20$as$u20$rustls..crypto..signer..Signer$GT$::sign::h39452a5eb77a90fc)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_PKEY_CTX_set_rsa_pss_saltlen
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_digest_sig::hf138b2418df46f53)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$rustls..crypto..aws_lc_rs..sign..RsaSigner$u20$as$u20$rustls..crypto..signer..Signer$GT$::sign::h39452a5eb77a90fc)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_DigestVerify
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced 1 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_MD_CTX_cleanup
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..ec..signature..EcdsaVerificationAlgorithm$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::h33057cb83ecc8297)
          >>> referenced 23 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_PKEY_free
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$aws_lc_rs..rsa..signature..RsaParameters$u20$as$u20$aws_lc_rs..signature..VerificationAlgorithm$GT$::verify_sig::hdb5e52503c7f3495)
          >>> referenced 76 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_RSA_public_key_from_bytes
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::rsa::key::parse_rsa_public_key::h0df716eeb8b5be18)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_PKEY_new
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::rsa::key::parse_rsa_public_key::h0df716eeb8b5be18)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::ec::encoding::sec1::parse_sec1_public_point::h57ad8159bd564b26)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$rustls..crypto..aws_lc_rs..AwsLcRs$u20$as$u20$rustls..crypto..KeyProvider$GT$::load_private_key::h048862801c6bc91d)
          >>> referenced 1 more times
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_PKEY_assign_RSA
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::rsa::key::parse_rsa_public_key::h0df716eeb8b5be18)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$rustls..crypto..aws_lc_rs..AwsLcRs$u20$as$u20$rustls..crypto..KeyProvider$GT$::load_private_key::h048862801c6bc91d)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_RSA_free
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::rsa::key::parse_rsa_public_key::h0df716eeb8b5be18)
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(_$LT$rustls..crypto..aws_lc_rs..AwsLcRs$u20$as$u20$rustls..crypto..KeyProvider$GT$::load_private_key::h048862801c6bc91d)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha1
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha512
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha256
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha384
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha3_384
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha224
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha512_256
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: undefined symbol: aws_lc_0_41_0_EVP_sha3_256
          >>> referenced by dalfox.bb595408c8d194d-cgu.0
          >>>               /home/linuxmobile/.cache/paru/clone/dalfox/src/dalfox-3.0.1/target/release/deps/dalfox-8d2397260a65e26a.dalfox.bb595408c8d194d-cgu.0.rcgu.o:(aws_lc_rs::digest::match_digest_type::heb73095b4e1fd0b1)
          
          ld.lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
          collect2: error: ld returned 1 exit status
          

error: could not compile `dalfox` (bin "dalfox") due to 1 previous error
==> ERROR: A failure occurred in build().
    Aborting...
error: failed to build 'dalfox-3.0.1-1': 
error: packages failed to build: dalfox-3.0.1-1
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -13,9 +13,15 @@ repository = "https://github.com/hahwul/dalfox"
 clap = { version = "4.0", features = ["derive", "env"] }
 url = "2.0"
 reqwest = { version = "0.13", default-features = false, features = [
-    "rustls",
+    "rustls-no-provider",
     "socks", "multipart", "json", "form", "gzip", "deflate", "brotli"
 ] }
+# Pin rustls to the `ring` crypto backend instead of the default `aws-lc-rs`.
+# aws-lc-sys ships C/asm that fails to link on several distros (AUR, musl,
+# etc. — see #1061), whereas ring's prebuilt assembly is portable. reqwest's
+# `rustls-no-provider` feature leaves the crypto provider unset, so we install
+# ring's as the process default (see `dalfox::ensure_crypto_provider`).
+rustls = { version = "0.23", default-features = false, features = ["ring", "std", "tls12", "logging"] }
 scraper = "0.27"
 tokio = { version = "1", features = ["full"] }
 serde = { version = "1.0", features = ["derive"] }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -79,6 +79,26 @@ pub static WAF_CONSECUTIVE_BLOCKS: std::sync::atomic::AtomicU32 =
     std::sync::atomic::AtomicU32::new(0);
 pub static NO_COLOR: AtomicBool = AtomicBool::new(false);
 
+/// Install the process-wide rustls crypto provider (ring).
+///
+/// reqwest is built with the `rustls-no-provider` feature so it bundles no
+/// crypto backend; we supply ring's instead of the default aws-lc-rs because
+/// aws-lc-sys fails to link on several distros (AUR, musl, …; see #1061).
+/// reqwest resolves the provider via `CryptoProvider::get_default()` when a
+/// `Client` is built, so this must run before the first client is constructed
+/// — it is called from `main()` and from every pooled client builder.
+///
+/// Idempotent and thread-safe: the first caller installs the provider, the
+/// rest are no-ops. If a downstream library consumer already installed a
+/// provider of their own, `install_default` returns `Err` and we leave their
+/// choice untouched.
+pub fn ensure_crypto_provider() {
+    static INSTALL: std::sync::Once = std::sync::Once::new();
+    INSTALL.call_once(|| {
+        let _ = rustls::crypto::ring::default_provider().install_default();
+    });
+}
+
 tokio::task_local! {
     /// Per-scan request counter, set by the MCP runner so concurrent scans
     /// don't pollute each other's progress tallies. Callers use

--- a/src/main.rs
+++ b/src/main.rs
@@ -77,6 +77,11 @@ use dalfox::utils::fs::read_bounded;
 
 #[tokio::main]
 async fn main() {
+    // Install the rustls crypto provider (ring) before anything builds a
+    // reqwest Client. reqwest uses `rustls-no-provider`, so without this the
+    // first Client::build() panics with "no crypto provider configured".
+    dalfox::ensure_crypto_provider();
+
     // Exit cleanly when a downstream consumer (e.g. `head`, `grep -q`) closes
     // the pipe. Rust ignores SIGPIPE by default, so the next `println!` panics
     // inside the stdio shim with `failed printing to stdout: Broken pipe` and

--- a/src/payload/remote.rs
+++ b/src/payload/remote.rs
@@ -93,6 +93,7 @@ pub struct RemoteFetchOptions {
 
 /// Build a reqwest Client with the given remote fetch options (timeout, proxy).
 fn build_remote_client(opts: &RemoteFetchOptions) -> Result<Client, Box<dyn std::error::Error>> {
+    crate::ensure_crypto_provider();
     let mut builder = Client::builder()
         .timeout(Duration::from_secs(
             opts.timeout_secs.unwrap_or(DEFAULT_TIMEOUT_SECS),
@@ -174,6 +175,7 @@ pub async fn init_remote_payloads(providers: &[String]) -> Result<(), Box<dyn st
         return Ok(());
     }
 
+    crate::ensure_crypto_provider();
     let client = Client::builder()
         .timeout(Duration::from_secs(DEFAULT_TIMEOUT_SECS))
         .danger_accept_invalid_certs(true)
@@ -209,6 +211,7 @@ pub async fn init_remote_wordlists(providers: &[String]) -> Result<(), Box<dyn s
         return Ok(());
     }
 
+    crate::ensure_crypto_provider();
     let client = Client::builder()
         .timeout(Duration::from_secs(DEFAULT_TIMEOUT_SECS))
         .danger_accept_invalid_certs(true)

--- a/src/scanning/light_verify/tests.rs
+++ b/src/scanning/light_verify/tests.rs
@@ -13,6 +13,15 @@ use std::collections::HashMap;
 use std::net::{Ipv4Addr, SocketAddr};
 use tokio::time::{Duration, sleep};
 
+/// reqwest is built with `rustls-no-provider`, so the ring crypto provider
+/// must be installed before the first `Client::build()` or it panics.
+/// Production installs it in `main()` / the pooled builders; these unit tests
+/// construct bare clients, so route them through this helper.
+fn test_client() -> Client {
+    crate::ensure_crypto_provider();
+    Client::new()
+}
+
 #[derive(Clone)]
 struct TestState {
     class_marker: String,
@@ -155,7 +164,7 @@ async fn test_verify_dom_xss_light_raw_reflection_without_marker_evidence() {
     let target = make_target(addr, "/reflect", None, None);
     let param = make_param(Location::Query, "q");
     let payload = "<script>alert(1)</script>";
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, note) =
         verify_dom_xss_light_with_client(&client, &target, &param, payload).await;
@@ -175,7 +184,7 @@ async fn test_verify_dom_xss_light_marker_element_present_without_payload() {
     let target = make_target(addr, "/marker-only", None, None);
     let param = make_param(Location::Query, "q");
     let payload = format!("<div class=\"{}\">injected</div>", marker);
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, note) =
         verify_dom_xss_light_with_client(&client, &target, &param, &payload).await;
@@ -192,7 +201,7 @@ async fn test_verify_dom_xss_light_sets_csp_hint_when_inline_handlers_blocked() 
     let target = make_target(addr, "/csp", None, None);
     let param = make_param(Location::Query, "q");
     let payload = format!("payload-{}", marker);
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, note) =
         verify_dom_xss_light_with_client(&client, &target, &param, &payload).await;
@@ -214,7 +223,7 @@ async fn test_verify_dom_xss_light_returns_none_response_on_request_failure() {
     target.timeout = 1;
     let param = make_param(Location::Query, "q");
     let payload = "x";
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, note) =
         verify_dom_xss_light_with_client(&client, &target, &param, payload).await;
@@ -231,6 +240,7 @@ async fn test_verify_dom_xss_light_redirection() {
     let target = make_target(addr, "/redirect", None, None);
     let param = make_param(Location::Query, "q");
     let payload = format!("<img class=\"{}\" src=x onerror=1>", marker);
+    crate::ensure_crypto_provider();
     let client = Client::builder()
         .redirect(reqwest::redirect::Policy::none())
         .build()
@@ -251,7 +261,7 @@ async fn test_verify_dom_xss_light_location_header() {
     let target = make_target(addr, "/header", None, None);
     let param = make_param(Location::Header, "X-XSS");
     let payload = format!("<img class=\"{}\" src=x onerror=1>", marker);
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, note) =
         verify_dom_xss_light_with_client(&client, &target, &param, &payload).await;
@@ -268,7 +278,7 @@ async fn test_verify_dom_xss_light_location_body() {
     let target = make_target(addr, "/body", Some("POST"), Some("q=seed"));
     let param = make_param(Location::Body, "q");
     let payload = format!("<img class=\"{}\" src=x onerror=1>", marker);
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, _note) =
         verify_dom_xss_light_with_client(&client, &target, &param, &payload).await;
@@ -284,7 +294,7 @@ async fn test_verify_dom_xss_light_location_json_body() {
     let target = make_target(addr, "/json", Some("POST"), Some("{\"q\":\"seed\"}"));
     let param = make_param(Location::JsonBody, "q");
     let payload = format!("<img class=\"{}\" src=x onerror=1>", marker);
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, _note) =
         verify_dom_xss_light_with_client(&client, &target, &param, &payload).await;
@@ -300,7 +310,7 @@ async fn test_verify_dom_xss_light_location_multipart_body() {
     let target = make_target(addr, "/multipart", Some("POST"), Some("q=seed"));
     let param = make_param(Location::MultipartBody, "q");
     let payload = format!("<img class=\"{}\" src=x onerror=1>", marker);
-    let client = Client::new();
+    let client = test_client();
 
     let (verified, response, _note) =
         verify_dom_xss_light_with_client(&client, &target, &param, &payload).await;

--- a/src/target_parser/mod.rs
+++ b/src/target_parser/mod.rs
@@ -89,6 +89,9 @@ impl Target {
     }
 
     pub fn build_client(&self) -> Result<Client, Box<dyn std::error::Error>> {
+        // Library consumers may build clients without going through `main()`,
+        // so make sure the ring crypto provider is installed first.
+        crate::ensure_crypto_provider();
         let key = self.client_cache_key();
         // Fast path: return a cached Client if one matches the key.
         if let Ok(guard) = client_cache().lock()

--- a/src/utils/http/tests.rs
+++ b/src/utils/http/tests.rs
@@ -143,6 +143,7 @@ fn test_build_preflight_request_get_sets_range_and_headers() {
     target.user_agent = Some("Dalfox-Test-UA".to_string());
     target.cookies = vec![("sid".to_string(), "abc".to_string())];
 
+    crate::ensure_crypto_provider();
     let client = reqwest::Client::new();
     let req = build_preflight_request(&client, &target, false, Some(128))
         .build()
@@ -179,6 +180,7 @@ fn test_build_preflight_request_get_sets_range_and_headers() {
 #[test]
 fn test_build_preflight_request_head_ignores_range() {
     let target = parse_target("https://example.com/path").unwrap();
+    crate::ensure_crypto_provider();
     let client = reqwest::Client::new();
     let req = build_preflight_request(&client, &target, true, Some(128))
         .build()

--- a/tests/functional/xss_mock_server.rs
+++ b/tests/functional/xss_mock_server.rs
@@ -1642,6 +1642,7 @@ async fn test_header_reflection_v2() {
         let header_name_lc = header_name.to_ascii_lowercase();
 
         // Simple functional check: server reflects the header value (transformed)
+        dalfox::ensure_crypto_provider();
         let client = reqwest::Client::new();
         let url = format!("http://{}:{}/header/{}", addr.ip(), addr.port(), case_id);
         let resp = client
@@ -1762,6 +1763,7 @@ async fn test_body_reflection_v2() {
         let param_name = case.param_name.as_deref().unwrap_or("query");
 
         // Simple functional check: server reflects the body param (transformed)
+        dalfox::ensure_crypto_provider();
         let client = reqwest::Client::new();
         let url = format!("http://{}:{}/body/{}", addr.ip(), addr.port(), case_id);
         let form = [(param_name.to_string(), "seed".to_string())];
@@ -1789,6 +1791,7 @@ async fn test_header_reflection_diverse_xss_contexts_v2() {
     // Representative header contexts: raw/script/attribute/js-string/style/encoded.
     let case_ids = vec![1, 9, 23, 32, 35, 41, 46, 49];
     let mut failed = Vec::new();
+    dalfox::ensure_crypto_provider();
     let client = reqwest::Client::new();
 
     for case_id in case_ids {
@@ -1834,6 +1837,7 @@ async fn test_header_reflection_advanced_xss_coverage_v2() {
     let case_ids = vec![42, 43, 45, 47, 48, 50, 34, 40];
     let mut reflected_count = 0usize;
     let mut mismatches = Vec::new();
+    dalfox::ensure_crypto_provider();
     let client = reqwest::Client::new();
 
     for case_id in case_ids {
@@ -2094,6 +2098,7 @@ async fn test_body_reflection_diverse_xss_contexts_v2() {
 
     let case_ids = vec![1, 9, 11, 14, 18, 21, 26, 31, 33];
     let mut failed = Vec::new();
+    dalfox::ensure_crypto_provider();
     let client = reqwest::Client::new();
 
     for case_id in case_ids {
@@ -2139,6 +2144,7 @@ async fn test_body_reflection_advanced_xss_coverage_v2() {
     let case_ids = vec![27, 28, 29, 30, 32, 34, 35, 36];
     let mut reflected_count = 0usize;
     let mut mismatches = Vec::new();
+    dalfox::ensure_crypto_provider();
     let client = reqwest::Client::new();
 
     for case_id in case_ids {

